### PR TITLE
[clang-tidy] Update llvmlibc-implementation-in-namespace to new rules

### PR DESCRIPTION
**Pull request by gchatelet**
This is the implementation of step 3 of
https://discourse.llvm.org/t/rfc-customizable-namespace-to-allow-testing-the-libc-when-the-system-libc-is-also-llvms-libc/73079.

### CODE REVIEW

File: clang-tools-extra/clang-tidy/llvmlibc/ImplementationInNamespaceCheck.cpp [hunk 1]
```diff
@@ -29,16 +31,19 @@ void ImplementationInNamespaceCheck::check(
   if (!Result.SourceManager->isInMainFile(MatchedDecl->getLocation()))
     return;
 
-  if (const auto *NS = dyn_cast<NamespaceDecl>(MatchedDecl)) {
-    if (NS->getName() != RequiredNamespace) {
-      diag(NS->getLocation(), "'%0' needs to be the outermost namespace")
-          << RequiredNamespace;
-    }
+  if (auto *NS = dyn_cast<NamespaceDecl>(MatchedDecl)) {
```
Comment (EugeneZelenko) [inline clang-tools-extra/clang-tidy/llvmlibc/ImplementationInNamespaceCheck.cpp:34]: `const auto *`

**Review by carlosgalvezp — COMMENTED** [review]
Should this change be reflected in the Release Notes?

**Comment by gchatelet** [comment]
> Should this change be reflected in the Release Notes?

Where would that be? `clang-tools-extra/docs/ReleaseNotes.rst`?
I can't find what the current release notes look like for head.

**Comment by PiotrZSL** [comment]
> Where would that be? `clang-tools-extra/docs/ReleaseNotes.rst`? I can't find what the current release notes look like for head.

Yes, it's deployed here: https://clang.llvm.org/extra/ReleaseNotes.html

File: clang-tools-extra/clang-tidy/llvmlibc/ImplementationInNamespaceCheck.cpp [hunk 2]
```diff
@@ -30,15 +32,18 @@ void ImplementationInNamespaceCheck::check(
...
   if (const auto *NS = dyn_cast<NamespaceDecl>(MatchedDecl)) {
-    if (NS->getName() != RequiredNamespace) {
-      diag(NS->getLocation(), "'%0' needs to be the outermost namespace")
-          << RequiredNamespace;
-    }
+    if (!Result.SourceManager->isMacroBodyExpansion(NS->getLocation()))
+      diag(NS->getLocation(),
+           "the outermost namespace should be the '%0' macro")
+          << RequiredNamespaceMacroName;
+    else if (!NS->getName().starts_with(RequiredNamespaceStart))
```
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/llvmlibc/ImplementationInNamespaceCheck.cpp:39]: getName will assert if Namespace got not identifier. What about anonymous namespaces ? Or declarations without name ?

File: clang-tools-extra/test/clang-tidy/checkers/llvmlibc/implementation-in-namespace.cpp [hunk 3]
```diff
@@ -3,18 +3,18 @@
...
-// CHECK-MESSAGES: :[[@LINE-1]]:13: warning: declaration must be declared within the '__llvm_libc' namespace
+// CHECK-MESSAGES: :[[@LINE-1]]:13: warning: declaration must be declared within a namespace starting with '__llvm_libc'
 void funcF() {}
-// CHECK-MESSAGES: :[[@LINE-1]]:6: warning: declaration must be declared within the '__llvm_libc' namespace
+// CHECK-MESSAGES: :[[@LINE-1]]:6: warning: declaration must be declared within a namespace starting with '__llvm_libc'
 
 namespace namespaceG {
-// CHECK-MESSAGES: :[[@LINE-1]]:11: warning: '__llvm_libc' needs to be the outermost namespace
+// CHECK-MESSAGES: :[[@LINE-1]]:11: warning: the outermost namespace should be the 'LIBC_NAMESPACE' macro
 namespace __llvm_libc{
```
Comment (PiotrZSL) [inline clang-tools-extra/test/clang-tidy/checkers/llvmlibc/implementation-in-namespace.cpp:18]: this looks to be invalid as its namespaceG::__llvm_libc and it's not "the outermost namespace should be the 'LIBC_NAMESPACE'"

File: clang-tools-extra/test/clang-tidy/checkers/llvmlibc/implementation-in-namespace.cpp [hunk 4]
```diff
@@ -3,18 +3,18 @@
...
+// CHECK-MESSAGES: :[[@LINE-1]]:8: warning: declaration must be declared within a namespace starting with '__llvm_libc'
 char *VarD = MACRO_A;
-// CHECK-MESSAGES: :[[@LINE-1]]:7: warning: declaration must be declared within the '__llvm_libc' namespace
+// CHECK-MESSAGES: :[[@LINE-1]]:7: warning: declaration must be declared within a namespace starting with '__llvm_libc'
 typedef int typeE;
-// CHECK-MESSAGES: :[[@LINE-1]]:13: warning: declaration must be declared within the '__llvm_libc' namespace
+// CHECK-MESSAGES: :[[@LINE-1]]:13: warning: declaration must be declared within a namespace starting with '__llvm_libc'
 void funcF() {}
-// CHECK-MESSAGES: :[[@LINE-1]]:6: warning: declaration must be declared within the '__llvm_libc' namespace
+// CHECK-MESSAGES: :[[@LINE-1]]:6: warning: declaration must be declared within a namespace starting with '__llvm_libc'
```
Comment (PiotrZSL) [inline clang-tools-extra/test/clang-tidy/checkers/llvmlibc/implementation-in-namespace.cpp:15]: add test:

```
namespace outer_most {
 class A {}; // i think it wont provide warning 'declaration must be declared within a namespace starting with ' to A class
}
```

File: clang-tools-extra/clang-tidy/llvmlibc/ImplementationInNamespaceCheck.cpp [hunk 5]
```diff
@@ -14,7 +14,9 @@ using namespace clang::ast_matchers;
 
 namespace clang::tidy::llvm_libc {
 
-const static StringRef RequiredNamespace = "__llvm_libc";
+const static StringRef RequiredNamespaceStart = "__llvm_libc";
+const static StringRef RequiredNamespaceMacroName = "LIBC_NAMESPACE";
+
 void ImplementationInNamespaceCheck::registerMatchers(MatchFinder *Finder) {
   Finder->addMatcher(
       decl(hasParent(translationUnitDecl()), unless(linkageSpecDecl()))
```
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/llvmlibc/ImplementationInNamespaceCheck.cpp:22]: Maybe: hasParent -> hasDeclContext(translationUnitDecl())

**Review by PiotrZSL — APPROVED** [review]
As a single small change to make hardcoded namespaces configurable looks fine.
I added some comments related to overall issues in this check.
Fell free to fix them or ignore them.

File: clang-tools-extra/clang-tidy/llvmlibc/ImplementationInNamespaceCheck.cpp [hunk 6]
```diff
@@ -30,15 +32,18 @@ void ImplementationInNamespaceCheck::check(
...
+    if (!Result.SourceManager->isMacroBodyExpansion(NS->getLocation()))
+      diag(NS->getLocation(),
+           "the outermost namespace should be the '%0' macro")
+          << RequiredNamespaceMacroName;
+    else if (!NS->getName().starts_with(RequiredNamespaceStart))
+      diag(NS->getLocation(), "the outermost namespace should start with '%0'")
+          << RequiredNamespaceStart;
     return;
   }
   diag(MatchedDecl->getLocation(),
```
[the remark this answers is not in the document]
Comment (gchatelet) [reply clang-tools-extra/clang-tidy/llvmlibc/ImplementationInNamespaceCheck.cpp:44]: This is the case where we are not in any namespace but the message is misleading. Thx for pointing this out. I'll fix it shortly.

File: clang-tools-extra/clang-tidy/llvmlibc/ImplementationInNamespaceCheck.cpp [hunk 7]
```diff
@@ -30,15 +32,18 @@ void ImplementationInNamespaceCheck::check(
     return;
```
[the remark this answers is not in the document]
Comment (gchatelet) [reply clang-tools-extra/clang-tidy/llvmlibc/ImplementationInNamespaceCheck.cpp:40]: Done

File: clang-tools-extra/test/clang-tidy/checkers/llvmlibc/implementation-in-namespace.cpp (hunk 4 above)
Comment (gchatelet) [reply clang-tools-extra/test/clang-tidy/checkers/llvmlibc/implementation-in-namespace.cpp:26]: done

**Comment by gchatelet** [comment]
> As a single small change to make hardcoded namespaces configurable looks fine. I added some comments related to overall issues in this check. Fell free to fix them or ignore them.

Thx for the review! I had to restructure the code a bit to accommodate for anonymous namespaces (they generate an implict UsingDirectiveDecl that was triggering the warning twice).

Let me know what you think.

File: clang-tools-extra/clang-tidy/llvmlibc/ImplementationInNamespaceCheck.cpp [hunk 8]
```diff
@@ -18,32 +18,32 @@ const static StringRef RequiredNamespaceStart = "__llvm_libc";
 const static StringRef RequiredNamespaceMacroName = "LIBC_NAMESPACE";
 
 void ImplementationInNamespaceCheck::registerMatchers(MatchFinder *Finder) {
-  Finder->addMatcher(
-      decl(hasParent(translationUnitDecl()), unless(linkageSpecDecl()))
-          .bind("child_of_translation_unit"),
-      this);
+  Finder->addMatcher(decl(isExpansionInMainFile(),
```
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/llvmlibc/ImplementationInNamespaceCheck.cpp:21]: On a hand, we could try oposite:
```
  Finder->addMatcher(translationUnitDecl(forEach(decl(isExpansionInMainFile(),
                          unless(linkageSpecDecl()))
                         .bind("child_of_translation_unit"))),
                     this);
```

File: clang-tools-extra/clang-tidy/llvmlibc/ImplementationInNamespaceCheck.cpp [hunk 9]
```diff
@@ -18,32 +18,32 @@ const static StringRef RequiredNamespaceStart = "__llvm_libc";
...
-  if (!Result.SourceManager->isInMainFile(MatchedDecl->getLocation()))
-    return;
 
   if (const auto *NS = dyn_cast<NamespaceDecl>(MatchedDecl)) {
     if (!Result.SourceManager->isMacroBodyExpansion(NS->getLocation()))
       diag(NS->getLocation(),
            "the outermost namespace should be the '%0' macro")
           << RequiredNamespaceMacroName;
-    else if (!NS->getName().starts_with(RequiredNamespaceStart))
+    else if (NS->isAnonymousNamespace() ||
```
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/llvmlibc/ImplementationInNamespaceCheck.cpp]: shouldn't we exclude anonymous namespaces at all ? or they not allowed ?

**Review by PiotrZSL — APPROVED** [review]
LGTM

File: clang-tools-extra/clang-tidy/llvmlibc/ImplementationInNamespaceCheck.cpp (hunk 9 above)
Comment (gchatelet) [reply clang-tools-extra/clang-tidy/llvmlibc/ImplementationInNamespaceCheck.cpp]: Yeah anonymous namespaces are not allowed.

**Pull request merged**